draft-ietf-tls-tls13.md

```diff
@@ -4073,10 +4073,10 @@ a non-zero octet in the cleartext, it MUST terminate the
 connection with an "unexpected_message" alert.
 
 The presence of padding does not change the overall record size limitations
-- the full encoded TLSInnerPlaintext MUST not exceed 2^14 octets. If the maximum
-fragment length is reduced, as for example by the max_fragment_length extension
-from [RFC6066], then the reduced limit applies to the full plaintext,
-including the padding.
+- the full encoded TLSInnerPlaintext MUST not exceed 2^14 + 1 octets. If the
+maximum fragment length is reduced, as for example by the max_fragment_length
+extension from [RFC6066], then the reduced limit applies to the full plaintext,
+including the content type and padding.
```

Comment:
Should max_fragment_length also exclude the content type? It seems sort of weird that, if you previously believed max_send_fragment was 2^14, you don't need version-specific logic, but if you believed it was 2^9, 2^10, 2^11, or 2^12 by way of the extension, you need version-specific logic again.

Reply:
Given the use case (constrained devices with limited memory) I think that it should include content type and padding. The limit seems to exist to reduce the memory requirement for the AEAD plaintext rather than the application data fragment.

The proposed text also matches with @martinthomson's https://tools.ietf.org/html/draft-ietf-tls-record-limit-01#section-4
>   This value is the length of the plaintext of a protected record.  The
>   value includes the content type and padding added in TLS 1.3 (that
>   is, the complete length of TLSInnerPlaintext).

Reply:
I see the value of @davidben's point about version-specific logic, but I think the point that @Lekensteyn makes about padding is pretty persuasive. ISTM that there are two sensible resolutions here:

- Move the record limit draft to talk about ciphertext, thus capping the maximum memory consumption under the theory that the worst-case is no padding. And then the plaintext limit would be "whatever fits in here minus 256, which is our worst-case expansion overhead"
- Have this limit talk about the total size of the plaintext, including CT and padding.

-Ekr

Reply:
I prefer the latter because it is more directly coupled to how records are produced.  

I also think that padding needs to be included (or enforcing the limit risks exposing when padding is in use).  In the simple case, you can simply say that the space available for plaintext is one less than the size limit (the version specific logic is `if (tls13) --limit;`, which is manageable).

Talking about ciphertext means that you have to include knowledge of the ciphersuite in your calculations.  In TLS 1.3, that's relatively simple, but for CBC modes in other versions, it's terrible.

Reply:
I would also prefer the second option. The first option would "waste" some space by enforcing a plaintext limit equal to the worst-case overhead.